File: backend/src/main/java/com/backend/petplace/global/response/ApiResponse.java
package com.backend.petplace.global.response;

import lombok.AllArgsConstructor;
import lombok.Getter;

@AllArgsConstructor
@Getter
public class ApiResponse<T> {

  private String code;
  private String message;
  private T data;

  public static <T> ApiResponse<T> success(T data) {
    return new ApiResponse<>(ResponseCode.OK.getCode(), ResponseCode.OK.getMessage(), data);
  }

  public static <T> ApiResponse<T> success() {
    return new ApiResponse<>(ResponseCode.OK.getCode(), ResponseCode.OK.getMessage(), null);
  }

  public static <T> ApiResponse<T> create() {
    return new ApiResponse<>(ResponseCode.CREATED.getCode(), ResponseCode.CREATED.getMessage(), null);
  }

  public static <T> ApiResponse<T> create(T data) {
    return new ApiResponse<>(ResponseCode.CREATED.getCode(), ResponseCode.CREATED.getMessage(), data);
  }

  public static <T> ApiResponse<T> error(ResponseCode code) {
    return new ApiResponse<>(code.getCode(), code.getMessage(), null);
  }

  public static <T> ApiResponse<T> error(ErrorCode code) {
    return new ApiResponse<>(code.getCode(), code.getMessage(), null);
  }
}

File: backend/src/main/java/com/backend/petplace/global/response/ApiResponse.kt
package com.backend.petplace.global.response


data class ApiResponse<T> (
    val code: String,
    val message: String,
    val data: T?
) {
    companion object {

        @JvmStatic
        fun <T> success(data: T?): ApiResponse<T?> {
            return ApiResponse(ResponseCode.OK.code, ResponseCode.OK.message, data)
        }

        @JvmStatic
        fun <T> success(): ApiResponse<T?> {
            return ApiResponse(ResponseCode.OK.code, ResponseCode.OK.message, null)
        }

        @JvmStatic
        fun <T> create(): ApiResponse<T?> {
            return ApiResponse(ResponseCode.CREATED.code,ResponseCode.CREATED.message,null)
        }

        @JvmStatic
        fun <T> create(data: T?): ApiResponse<T?> {
            return ApiResponse(ResponseCode.CREATED.code, ResponseCode.CREATED.message, data)
        }

        @JvmStatic
        fun <T> error(code: ResponseCode): ApiResponse<T?> {
            return ApiResponse(code.code, code.message, null)
        }

        @JvmStatic
        fun <T> error(code: ErrorCode): ApiResponse<T?> {
            return ApiResponse(code.code, code.message, null)
        }
    }
}